File: app/assets/javascripts/domain/study/SpecimenType.js
/**
 * @author Nelson Loyola <loyola@ualberta.ca>
 * @copyright 2015 Canadian BioSample Repository (CBSR)
 */
define(function () {
  'use strict';

  /**
   * A standardized set of classifications that describe <em>what</em> a {@link domain.participant.Specimen
   * Specimen} is. Potential examples include: urine, whole blood, plasma, nail, protein, etc.
   *
   * @enum {string}
   * @memberOf domain.studies
   */
  var SpecimenType = {
    BUFFY_COAT:                   'Buffy coat',
    CDPA_PLASMA:                  'CDPA Plasma',
    CENTRIFUGED_URINE:            'Centrifuged Urine',
    CORD_BLOOD_MONONUCLEAR_CELLS: 'Cord Blood Mononuclear Cells',
    DNA_BLOOD:                    'DNA (Blood)',
    DNA_WHITE_BLOOD_CELLS:        'DNA (White blood cells)',
    DESCENDING_COLON:             'Descending Colon',
    DUODENUM:                     'Duodenum',
    FILTERED_URINE:               'Filtered Urine',
    FINGER_NAILS:                 'Finger Nails',
    HAIR:                         'Hair',
    HEMODIALYSATE:                'Hemodialysate',
    HEPARIN_BLOOD:                'Heparin Blood',
    ILEUM:                        'Ileum',
    JEJUNUM:                      'Jejunum',
    LITHIUM_HEPARIN_PLASMA:       'Lithium Heparin Plasma',
    MECONIUM_BABY:                'Meconium - BABY',
    NAN3_URINE:                   'NaN3 Urine',
    PAXGENE:                      'Paxgene' ,
    PERITONEAL_DIALYSATE:         'Peritoneal Dialysate',
    PLASMA_NA_HEPARIN_DAD:        'Plasma (Na Heparin) - DAD',
    PLASMA:                       'Plasma',
    PLATELET_FREE_PLASMA:         'Platelet free plasma',
    RNA:                          'RNA',
    RNA_CBMC:                     'RNA CBMC',
    RNA_LATER_BIOPSIES:           'RNAlater Biopsies',
    SERUM:                        'Serum',
    SODIUM_AZIDE_URINE:           'SodiumAzideUrine',
    SOURCE_WATER:                 'Source Water',
    STOOL:                        'Stool',
    TAP_WATER:                    'Tap Water',
    TRANSVERSE_COLON:             'Transverse Colon',
    URINE:                        'Urine',
    WHOLE_BLOOD:                  'Whole Blood',
    WHOLE_BLOOD_CPDA:             'Whole Blood CPDA',
    WHOLE_BLOOD_EDTA:             'Whole Blood EDTA',
    WHOLE_BLOOD_LI_HEP:           'Whole Blood Li Hep'
  };

  return SpecimenType;
});

File: app/org/biobank/domain/SpecimenType.scala
package org.biobank.domain

import play.api.libs.json._
import org.biobank.infrastructure.EnumUtils._

/** A standardized set of classifications that describe what a [[domain.participants.Specimen Specimen]] is.
  */
@SuppressWarnings(Array("org.wartremover.warts.Enumeration"))
object SpecimenType extends Enumeration {
  type SpecimenType = Value

  val BuffyCoat: Value                 = Value("Buffy coat")
  val CdpaPlasma: Value                = Value("CDPA Plasma")
  val CentrifugedUrine: Value          = Value("Centrifuged Urine")
  val CordBloodMononuclearCells: Value = Value("Cord Blood Mononuclear Cells")
  val DnaBlood: Value                  = Value("DNA (Blood)")
  val DnaWhiteBloodCells: Value        = Value("DNA (White blood cells)")
  val DescendingColon: Value           = Value("Descending Colon")
  val Duodenum: Value                  = Value("Duodenum")
  val FilteredUrine: Value             = Value("Filtered Urine")
  val FingerNails: Value               = Value("Finger Nails")
  val Hair: Value                      = Value("Hair")
  val Hemodialysate: Value             = Value("Hemodialysate")
  val HeparinBlood: Value              = Value("Heparin Blood")
  val Ileum: Value                     = Value("Ileum")
  val Jejunum: Value                   = Value("Jejunum")
  val LithiumHeparinPlasma: Value      = Value("Lithium Heparin Plasma")
  val MeconiumBaby: Value              = Value("Meconium - BABY")
  val NaN3Urine: Value                 = Value("NaN3 Urine")
  val Paxgene: Value                   = Value("Paxgene")
  val PeritonealDialysate: Value       = Value("Peritoneal Dialysate")
  val PlasmaNaHeparinDad: Value        = Value("Plasma (Na Heparin) - DAD")
  val Plasma: Value                    = Value("Plasma")
  val PlateletFreePlasma: Value        = Value("Platelet free plasma")
  val Rna: Value                       = Value("RNA")
  val RnaCBMC: Value                   = Value("RNA CBMC")
  val RnaLaterBiopsies: Value          = Value("RNAlater Biopsies")
  val Serum: Value                     = Value("Serum")
  val SodiumAzideUrine: Value          = Value("SodiumAzideUrine")
  val SourceWater: Value               = Value("Source Water")
  val Stool: Value                     = Value("Stool")
  val TapWater: Value                  = Value("Tap Water")
  val Urine: Value                     = Value("Urine")
  val TransverseColon: Value           = Value("Transverse Colon")
  val WholeBlood: Value                = Value("Whole Blood")
  val WholeBloodCpda: Value            = Value("Whole Blood CPDA")
  val WholeBloodEdta: Value            = Value("Whole Blood EDTA")
  val WholeBloodLiHep: Value           = Value("Whole Blood Li Hep")

  implicit val specimenTypeReads: Format[SpecimenType] = enumFormat(SpecimenType)
}
